idea/src/hev-timeline/hev-interpolateRotation.1
.TH hev-interpolateRotation 1 "November 2011"

.SH NAME
hev-interpolateRotation

.SH SYNOPSIS 
.B hev-interpolateRotation  [ --r0 \fIh p r\fB ] [ --r1 \fIh p r\fB ]  [ --absolute ] [ --startTime \fIt\fB ] [ --duration \fIt\fB ] [ --frameRate \fIr\fB ] [ --deltaRotation \fIa\fB ]

.SH OPTIONS

.IP "--r0 \fIh p r\fR"
specifies the initial rotation in Euler angles. If not specified, 0, 0, 0, is used.

.IP "--r1 \fIh p r\fR"
specifies the final rotation in Euler angles. If not specified, 0, 0, 0, is used.

.IP "--absolute"
specifies that the times used in the IRIS control commands be absolute
times. i.e. OSG reference times. The default is to use time values that are
relative to the time the control command was processed.

.IP "--startTime \fIt\fR"
specifies the initial time offset in seconds to be added to all IRIS control
commands.  The default value is 0.

.IP "--duration \fIt\fR"
specifies the number of seconds to rotate from r0 to r1. If not supplied,
duration will be set to 1.0 second.

.IP "--frameRate \fIr\fR"
specifies the number of times per second \fBhev-interpolateRotation\fR
will interpolate between the rotations.  If \fIframeRate\fR is not specified, 60 will be used.

.IP "--deltaRotation \fIa\fR"
specifies the amount that the navigation node's rotation value needs to
change before command is written. The amount is the angle from the
last written rotation.

If \fIdeltaRotation\fR is not specified, it defaults to .1 degrees.

If \fIdeltaRotation\fR is set to zero a command containing a rotation
value will be written \fIframeRate\fR times a second.

.SH DESCRIPTION

\fBhev-interpolateRotation\fR will write IRIS control commands that will
modify the navigation node's rotation to go from \fIr0\fR to
\fIr1\fR over the course of \fIduration\fR seconds, writing up to \fIframerate\fR commands
per second.

\fBhev-interpolateRotation\fR liberally comments the control commands it
generates. Like any IRIS control command file, the results can be edited with any text
editor.

.SH EXAMPLE

hev-interpolateRotation --r0 -45 -45 -45 --r1 45 45 45 --start 10 --duration 2.5 --frameRate 4  > /tmp/scoot.iris

will generate and store in the file /tmp/scoot.iris:

AFTER TIME 10.000000 NAV ORIENTATION -45.000000 -45.000000 -45.000000
.br
AFTER TIME 10.250000 NAV ORIENTATION -26.800400 -45.143848 -26.800400
.br
AFTER TIME 10.500000 NAV ORIENTATION -9.922147 -41.380863 -9.922147
.br
AFTER TIME 10.750000 NAV ORIENTATION 3.841135 -34.502022 3.841135
.br
AFTER TIME 11.000000 NAV ORIENTATION 14.417210 -25.500029 14.417210
.br
AFTER TIME 11.250000 NAV ORIENTATION 22.500000 -15.141508 22.500000
.br
AFTER TIME 11.500000 NAV ORIENTATION 28.811312 -3.920380 28.811312
.br
AFTER TIME 11.750000 NAV ORIENTATION 33.895695 7.860523 33.895695
.br
AFTER TIME 12.000000 NAV ORIENTATION 38.129871 20.013964 38.129871
.br
AFTER TIME 12.250000 NAV ORIENTATION 41.771431 32.420364 41.771431
.br
AFTER TIME 12.500000 NAV ORIENTATION 45.000000 45.000000 45.000000

Note that the first interpolated Euler angle generated might appear to be
quite different from the initial and final Euler angles; this is due to the fact
that there are several equivalent Euler angles that can specify the same
orientation.

.SH SEE ALSO

hev-timeline, hev-navigationToTimeline, hev-timelineToIris,
hev-interpolateScale, hev-interpolateTranslation, hev-readNav

.PP
John Kelso, kelso@nist.gov
.br
NIST High Performance Computing and Visualization Group
.br
November 2011
.br
\fBhttp://www.nist.gov/mcsd/hpcvg/\fR



idea/src/hev-timeline/hev-interpolateRotation.cpp
#include <stdio.h>
#include <iostream>
#include <stdlib.h>

#include <iris.h>

#include "utils.h"

void usage()
{
    fprintf(stderr,"Usage: hev-interpolateRotation [ --r0 h p r ] [ --r1 h p r ]  [ --absolute ] [ --startTime t ] [ --duration ] [ --frameRate ] [ --deltaRotation a ]\n") ;
}

////////////////////////////////////////////////////////////////////////
int main(int argc, char** argv)
{

    // send dtk messages to stdout
    dtkMsg.setFile(stderr) ;

    double r0[3] = {0, 0, 0} ;
    double r1[3] = {0, 0, 0} ;
    float startTime = 0 ;
    float duration = 1 ;
    float frameRate = -1 ;
    bool absolute = false ;
    double deltaRotation = .1 ;

    int c = 1 ;
    while (c < argc && argv[c][0] == '-' && argv[c][1] == '-')
    {
	if (c>=argc || iris::IsSubstring("--absolute",argv[c],3))
	{
	    c++ ;
	    absolute = true ;
	}
	else if (c>=argc || iris::IsSubstring("--startTime",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&startTime) || startTime<0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateRotation: invalid startTime value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--duration",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&duration) || duration<=0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateRotation: invalid duration value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--frameRate",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&frameRate) || frameRate<=0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateRotation: invalid frameRate value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--deltaRotation",argv[c],3))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToDouble(argv[c],&deltaRotation) || deltaRotation<0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateRotation: invalid deltaRotation value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || !strcmp("--r0",argv[c]))
	{
	    c++ ;
	    if (c>=argc-2 || !iris::StringToDouble(argv[c],r0) || !iris::StringToDouble(argv[c+1],r0+1) || !iris::StringToDouble(argv[c+2],r0+2)) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateRotation: invalid --r0 value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c+=3 ;
	}
	else if (c>=argc || !strcmp("--r1",argv[c]))
	{
	    c++ ;
	    if (c>=argc-2 || !iris::StringToDouble(argv[c],r1) || !iris::StringToDouble(argv[c+1],r1+1) || !iris::StringToDouble(argv[c+2],r1+2)) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateRotation: invalid --r0 value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c+=3 ;
	}
	else 
	{
	    dtkMsg.add(DTKMSG_ERROR, "hev-hev-interpolateRotation: unrecognized option: %s\n",argv[c]) ;
	    usage() ;
	    return 1 ;
	}
    }

#if 0
    // get the frame rate from the system if frameRate not passed as a parameter
    if (frameRate < 0) frameRate = getDefaultFrameRate() ;
#else
    if (frameRate < 0) frameRate = 60 ;
#endif
    if (frameRate < 0) frameRate = getDefaultFrameRate() ;

    std::string timeString ;
    if (absolute) 
    {
	timeString = "ABSTIME" ;
	printf("# time is relative to start of first graphics frame\n") ;
    }
    else 
    {
	timeString = "TIME" ;
	printf("# time is relative to when control commands are received\n") ;
    }
    
    printf("# rotating from r0=(%f, %f, %f) to r1=(%f, %f, %f) ; start time = %f, duration = %f, frame rate = %f\n",r0[0], r0[1], r0[2], r1[0], r1[1], r1[2], startTime, duration, frameRate) ;

    Posn p0, p1;
    p0.att = iris::EulerToQuat(r0[0], r0[1], r0[2]) ; ;
    p0.t = startTime ;
    p1.att = iris::EulerToQuat(r1[0], r1[1], r1[2]) ;
    p1.t = startTime + duration ;

    Interpolate interp(p0, p1) ;

    // always write the first
    printf("# first t = %f\n",p0.t) ;
    printf("AFTER %s %f NAV ORIENTATION %f %f %f\n",timeString.c_str(), p0.t, r0[0], r0[1], r0[2]) ;
    Posn lastPosn = p0 ;

    for (int i=1; i<frameRate*duration; i++)
    {
	double t = p0.t + double(i)/frameRate ;
	Posn interpPosn ;
	printf("# interpolated t = %f\n",t) ;
	if (!interp.at(t, &interpPosn))
	{
	    dtkMsg.add(DTKMSG_ERROR, "hev-interpolateRotation: interpolation error at t=%f\n",t) ;
	}
	else
	{
	    // only write if it changed from last time
	    if (movedEnoughRot(interpPosn.att, lastPosn.att, deltaRotation))
	    {
		lastPosn.att = interpPosn.att ;
		double h, p, r ;
		iris::QuatToEuler(interpPosn.att, &h, &p, &r) ;
		printf("AFTER %s %f NAV ORIENTATION %f %f %f\n",timeString.c_str(), t, h, p, r) ;
		fflush(stdout) ;
	    }
	}
    }
    // always write the last
    printf("# last t = %f\n",p1.t) ;
    printf("AFTER %s %f NAV ORIENTATION %f %f %f\n",timeString.c_str(), p1.t, r1[0], r1[1], r1[2]) ;
    return 0 ;
}
